[SPARK-16288][SQL] Implement inline table generating function

dongjoon-hyun opened the pull request:
## What changes were proposed in this pull request?

This PR implements `inline` table generating function.
## How was this patch tested?

Pass the Jenkins tests with new testcase.

dongjoon-hyun committed:
[SPARK-16288][SQL] Implement inline table generating function

SparkQA commented:
**[Test build #61479 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61479/consoleFull)** for PR 13976 at commit [`1add31a`](https://github.com/apache/spark/commit/1add31a1472ada2ac20e676873a4dc88d9c8393f).

SparkQA commented:
**[Test build #61479 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61479/consoleFull)** for PR 13976 at commit [`1add31a`](https://github.com/apache/spark/commit/1add31a1472ada2ac20e676873a4dc88d9c8393f).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Inline(child: Expression) extends UnaryExpression with Generator with CodegenFallback `

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61479/
Test FAILed.

dongjoon-hyun commented:
The failure seems to be irrelevant to this PR.
```
Exception encountered when attempting to run a suite with class name: org.apache.spark.sql.jdbc.JDBCWriteSuite *** ABORTED *** (886 milliseconds)
[info]   org.h2.jdbc.JdbcSQLException: Schema "TEST" already exists; SQL statement:
```

dongjoon-hyun commented:
Retest this please.

SparkQA commented:
**[Test build #61494 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61494/consoleFull)** for PR 13976 at commit [`1add31a`](https://github.com/apache/spark/commit/1add31a1472ada2ac20e676873a4dc88d9c8393f).

SparkQA commented:
**[Test build #61494 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61494/consoleFull)** for PR 13976 at commit [`1add31a`](https://github.com/apache/spark/commit/1add31a1472ada2ac20e676873a4dc88d9c8393f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Inline(child: Expression) extends UnaryExpression with Generator with CodegenFallback `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61494/
Test PASSed.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -149,3 +149,42 @@ case class Explode(child: Expression) extends UnaryExpression with Generator wit
+ * Explodes an array of structs into a table.
+ */
+@ExpressionDescription(
+  usage = "_FUNC_(a) - Explodes an array of structs into a table.")
```
give an example in extended?

dongjoon-hyun replied:
Sure.

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -149,3 +149,42 @@ case class Explode(child: Expression) extends UnaryExpression with Generator wit
+
+  override def elementSchema: StructType = child.dataType match {
+    case ArrayType(et : StructType, _) =>
+      et.fields.zipWithIndex.foldLeft(new StructType()) { case (schema, (f, i)) =>
```
everytime I see a foldLeft I feel it would be more clear if it's written in a more imperative style ... :)

```
var schema = new StructType()
for (i <- et.fields.indices) {
  schema = schema.add(...)
}
schema
```

dongjoon-hyun replied:
Oh, I see. I'll fix this and remember that. :)

rxin commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -149,3 +149,42 @@ case class Explode(child: Expression) extends UnaryExpression with Generator wit
+        Nil
+      } else {
+        val rows = new Array[InternalRow](inputArray.numElements())
+        inputArray.foreach(et, (i, e) => {
```
we don't need to materialize the array, do we? We can create an iterator to return the results.

dongjoon-hyun replied:
Oh, sure. I'll update this, too.

SparkQA commented:
**[Test build #61542 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61542/consoleFull)** for PR 13976 at commit [`e8d284c`](https://github.com/apache/spark/commit/e8d284c6ec8eb50f793f5c54d162acde7dfbcacd).

SparkQA commented:
**[Test build #61542 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61542/consoleFull)** for PR 13976 at commit [`e8d284c`](https://github.com/apache/spark/commit/e8d284c6ec8eb50f793f5c54d162acde7dfbcacd).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61542/
Test PASSed.

SparkQA commented:
**[Test build #61562 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61562/consoleFull)** for PR 13976 at commit [`be408ac`](https://github.com/apache/spark/commit/be408ac47a7bb5cc2dd7e074b742d7f8331cf4ae).

dongjoon-hyun commented:
Hi, @rxin .

I rebased and updated according to the previous `PosExplode` PR.
By the way, I didn't register into `sql/Dataset.scala` and python/R since I was confused at `map_keys/map_values` in https://github.com/apache/spark/pull/13967.

If you give an advice, I will proceed this PR (inline) and #13967 (map_keys/map_values) further.

rxin commented:
looks pretty good. I will let @cloud-fan review and do the merge.

dongjoon-hyun commented:
Yep. Thank you, @rxin .

SparkQA commented:
**[Test build #61562 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61562/consoleFull)** for PR 13976 at commit [`be408ac`](https://github.com/apache/spark/commit/be408ac47a7bb5cc2dd7e074b742d7f8331cf4ae).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61562/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,43 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+      for (f <- et.fields.zipWithIndex) {
+        schema = schema.add(s"col${f._2 + 1}", f._1.dataType, nullable = f._1.nullable)
+      }
+      schema
```
how about
```
StructType(et.fields.zipWithIndex.map {
  case (field, index) => StructField(s"col{$index + 1}", field.dataType, nullable = field.nullable)
})
```

rxin replied:
actually that's the best !

dongjoon-hyun replied:
Sure. Thanks, @cloud-fan !

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,43 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+    case ArrayType(et : StructType, _) =>
+      var schema = new StructType()
+      for (f <- et.fields.zipWithIndex) {
+        schema = schema.add(s"col${f._2 + 1}", f._1.dataType, nullable = f._1.nullable)
```
is it hive's rule to use `col1`, `col2` etc. instead of the field name of the struct?

dongjoon-hyun replied:
Oh, I overridded the `named_struct` case.
Sorry, I'll fix this. Actually, `f._1.name` was consistent.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,43 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+  }
+
+  override def eval(input: InternalRow): TraversableOnce[InternalRow] = child.dataType match {
+    case ArrayType(NullType, _) => Nil
```
how can this pass the type check?

dongjoon-hyun replied:
Actually, I added this to pass the `Seq.empty` testcase.
https://github.com/apache/spark/pull/13976/files#diff-6715134a4e95980149a7600ecb71674cR80

But, now, I think I added this without deep thought.

dongjoon-hyun replied:
I'm also wondering how this pass the type checking.

cloud-fan replied:
the expression evaluate test won't go through the type check process

dongjoon-hyun replied:
Indeed. I'll move the testcase into end-to-end testsuite.

cloud-fan replied:
We still need the expression-level unit tests...

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,43 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+        Nil
+      } else {
+        for (i <- 0 until inputArray.numElements())
+          yield InternalRow(inputArray.array(i).asInstanceOf[GenericInternalRow].values: _*)
```
isn't it just `inputArray.getStruct(i)`?

dongjoon-hyun replied:
Do you mean `getStruct(int ordinal, int numFields)` ?

cloud-fan replied:
yup

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala:
```diff
@@ -89,4 +89,11 @@ class GeneratorFunctionSuite extends QueryTest with SharedSQLContext {
+  test("single inline") {
+    val df = Seq((1, Seq(1, 2, 3))).toDF("a", "intList")
+    checkAnswer(
+      df.selectExpr("inline(array(struct(10, 100), struct(20, 200), struct(30, 300)))"),
```
we also need a test having a FROM clause

dongjoon-hyun replied:
Yep. I'll add that.

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratorExpressionSuite.scala:
```diff
@@ -68,4 +68,23 @@ class GeneratorExpressionSuite extends SparkFunSuite with ExpressionEvalHelper {
+      Seq(2, UTF8String.fromString("c")))
+
+    checkTuple(
+      Inline(CreateArray(Seq.empty)),
```
we can use `Literal.create(Array(), ArrayType(...))` to create a empty array with expected type.

SparkQA commented:
**[Test build #61615 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61615/consoleFull)** for PR 13976 at commit [`9382f64`](https://github.com/apache/spark/commit/9382f64a19c9671a679a75ce22b801aa32576da5).

dongjoon-hyun commented:
Thank you, @cloud-fan . :)   I've learn a lot in this PR again.
The followings are applied.
- Use `inputArray.getStruct(i, ncol)`
- Keep the original field name
- Fix elementSchema generation style
- Add a column-based test, `Array()` expression-level test, add empty row test.

SparkQA commented:
**[Test build #61615 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61615/consoleFull)** for PR 13976 at commit [`9382f64`](https://github.com/apache/spark/commit/9382f64a19c9671a679a75ce22b801aa32576da5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds the following public classes _(experimental)_:
  * `case class Inline(child: Expression) extends UnaryExpression with Generator with CodegenFallback `

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61615/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,42 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+
+  override def elementSchema: StructType = child.dataType match {
+    case ArrayType(et : StructType, _) =>
+      StructType(et.fields.zipWithIndex.map {
```
hmm, so it's just `et` now?

dongjoon-hyun replied:
Yep. Currently, our type checker ensures that homogeneous StructType array.

cloud-fan replied:
why not return `et` directly?

dongjoon-hyun replied:
Oh, my god. I was too naive, here.
Thank you!

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,42 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+
+  private lazy val ncol = elementSchema.fields.length
+
+  override def eval(input: InternalRow): TraversableOnce[InternalRow] = child.dataType match {
```
Why do we pattern match here?

dongjoon-hyun replied:
Oh, really it's useless. It was for NullType before. I'll remove this.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,42 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+      })
+  }
+
+  private lazy val ncol = elementSchema.fields.length
```
I'd like to name it `numFields`

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratorExpressionSuite.scala:
```diff
@@ -68,4 +69,23 @@ class GeneratorExpressionSuite extends SparkFunSuite with ExpressionEvalHelper {
+      Seq(2, UTF8String.fromString("c")))
+
+    checkTuple(
+      Inline(Literal.create(Array(), ArrayType(StructType(Seq(StructField("id1", LongType)))))),
```
we usually use `new StructType().add("id", LongType)` to create struct type

cloud-fan commented on sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/GeneratorExpressionSuite.scala:
```diff
@@ -68,4 +69,23 @@ class GeneratorExpressionSuite extends SparkFunSuite with ExpressionEvalHelper {
+
+  test("inline") {
+    val correct_answer = Seq(
+      Seq(0, UTF8String.fromString("a")),
```
we can create a row directly in test: call `create_row(...)`

cloud-fan replied:
and it can help us convert string to UTF8String

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala:
```diff
@@ -89,4 +91,30 @@ class GeneratorFunctionSuite extends QueryTest with SharedSQLContext {
       Row(3) :: Nil)
   }
+
+  test("inline with empty table or empty array") {
```
the test name is misleading: we do allow empty array, the problem is `array()` returns an array of null, which fails the type check.

dongjoon-hyun committed:
Remove redundant match and use create_row() in testsuites.

dongjoon-hyun commented:
Thank you, @cloud-fan .
I renovated ugly testcases with `create_row()`!

SparkQA commented:
**[Test build #61622 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61622/consoleFull)** for PR 13976 at commit [`c43a187`](https://github.com/apache/spark/commit/c43a1871cb1c1d1754712429c8f37a5e211d641f).

shaneknapp commented:
jenkins, test this please

SparkQA commented:
**[Test build #61624 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61624/consoleFull)** for PR 13976 at commit [`c43a187`](https://github.com/apache/spark/commit/c43a1871cb1c1d1754712429c8f37a5e211d641f).

dongjoon-hyun commented:
Thank you, @shaneknapp ! :)

shaneknapp commented:
no problem...  sorry i had to kill it, but i wanted to get a quick jenkins restart in.  :)

SparkQA commented:
**[Test build #61624 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61624/consoleFull)** for PR 13976 at commit [`c43a187`](https://github.com/apache/spark/commit/c43a1871cb1c1d1754712429c8f37a5e211d641f).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61624/
Test PASSed.

cloud-fan commented on sql/core/src/test/scala/org/apache/spark/sql/GeneratorFunctionSuite.scala:
```diff
@@ -89,4 +89,32 @@ class GeneratorFunctionSuite extends QueryTest with SharedSQLContext {
       Row(3) :: Nil)
   }
+
+  test("inline raises exception on empty array") {
```
`on array of null type`?

dongjoon-hyun replied:
Yep. That's more clear.

dongjoon-hyun committed:
Simplify elementSchema and update a testcase name.

dongjoon-hyun commented:
Thank you, @cloud-fan .
`elementSchema` is simplified and the testcase name is updated.

SparkQA commented:
**[Test build #61647 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61647/consoleFull)** for PR 13976 at commit [`ebd9cfa`](https://github.com/apache/spark/commit/ebd9cfa1dc4ed2c9ed28653ddb2c1a214ca19add).

SparkQA commented:
**[Test build #61647 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61647/consoleFull)** for PR 13976 at commit [`ebd9cfa`](https://github.com/apache/spark/commit/ebd9cfa1dc4ed2c9ed28653ddb2c1a214ca19add).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61647/
Test PASSed.

dongjoon-hyun commented:
Oh, I found a bug and am working on this.

dongjoon-hyun committed:
Add more testcases. Spark array type verification is more strict.

SparkQA commented:
**[Test build #61669 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61669/consoleFull)** for PR 13976 at commit [`31ffa75`](https://github.com/apache/spark/commit/31ffa758cfd5aa41851cb77a15b03da6d54e9198).

dongjoon-hyun commented:
Actually, that is not a bug, but I found that there exists a little difference between Spark and Hive with the following query.
```
SELECT inline(array(struct(a), struct(b))) FROM (SELECT 1 a, 2 b) T
```
In short, Spark does more strict type-checking, e.g., `[struct<a:int>, struct<b:int>]` is considered heterogeneous due to name difference.

I only add more tests to clarify the cases. We cannot touch that because it depends on many things.

The following query is a workaround which both Spark/Hive work.
```
SELECT inline(array(struct(a), named_struct('a', b))) FROM (SELECT 1 a, 2 b) T
```

dongjoon-hyun committed:
Remove redundant import statement.

SparkQA commented:
**[Test build #61670 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61670/consoleFull)** for PR 13976 at commit [`fed3ba2`](https://github.com/apache/spark/commit/fed3ba2bde5f82946c49fef5c06c85791400cea5).

SparkQA commented:
**[Test build #61669 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61669/consoleFull)** for PR 13976 at commit [`31ffa75`](https://github.com/apache/spark/commit/31ffa758cfd5aa41851cb77a15b03da6d54e9198).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61669/
Test PASSed.

SparkQA commented:
**[Test build #61670 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61670/consoleFull)** for PR 13976 at commit [`fed3ba2`](https://github.com/apache/spark/commit/fed3ba2bde5f82946c49fef5c06c85791400cea5).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61670/
Test PASSed.

dongjoon-hyun committed:
Fix detail example description.

SparkQA commented:
**[Test build #61676 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61676/consoleFull)** for PR 13976 at commit [`e260359`](https://github.com/apache/spark/commit/e26035968c73210dda38e82654fc335390fc6c1e).

SparkQA commented:
**[Test build #61676 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/61676/consoleFull)** for PR 13976 at commit [`e260359`](https://github.com/apache/spark/commit/e26035968c73210dda38e82654fc335390fc6c1e).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/61676/
Test PASSed.

cloud-fan commented on sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/generators.scala:
```diff
@@ -195,3 +195,38 @@ case class Explode(child: Expression) extends ExplodeBase(child, position = fals
+      Nil
+    } else {
+      for (i <- 0 until inputArray.numElements())
+        yield inputArray.getStruct(i, numFields)
```
I'm not sure how is the performance of `for-yield`, maybe it's safe to create an array manually and use while loop here?

dongjoon-hyun replied:
Thank you, @cloud-fan . By the way, for about this, @rxin gave me an advice at the first commit of this PR.
> we don't need to materialize the array, do we? We can create an iterator to return the results.

cloud-fan replied:
ah i see, `for-yield` returns an iterator.

cloud-fan commented:
merging to master, thanks!

dongjoon-hyun commented:
Thank you, @cloud-fan and @rxin ! :)
